services/job_runner.go

```diff
@@ -73,19 +77,25 @@ func (rm *jobRunner) resumeSleepingRuns() error {
 		return err
 	}
 	for _, run := range pendingRuns {
-		rm.store.RunChannel.Send(run.Result, nil)
+		rm.store.RunChannel.Send(run.ID, run.Result, nil)
 	}
 	return nil
 }
 
 func (rm *jobRunner) demultiplexRuns() {
+	rm.demultiplexStopperWg.Add(1)
+	defer rm.demultiplexStopperWg.Done()
 	for {
 		select {
 		case <-rm.done:
 			logger.Debug("JobRunner demultiplexing of job runs finished")
 			return
-		case rr := <-rm.store.RunChannel.Receive():
-			rm.channelForRun(rr.Input.JobRunID) <- rr
+		case rr, ok := <-rm.store.RunChannel.Receive():
+			if !ok {
+				logger.Error("JobRunner RunChannel closed, demultiplexing of job runs finished")
```

Comment:
If I would put a `panic` due to an invariant, I think it would be here. If this closes prematurely, we're in a world of hurt.